flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/adaptive/AdaptiveJoinOperatorGenerator.java

```diff
@@ -0,0 +1,184 @@
+
+/**
+ * Implementation class for {@link AdaptiveJoin}. It can selectively generate broadcast hash join,
+ * shuffle hash join or shuffle merge join operator based on actual conditions.
+ */
+public class AdaptiveJoinOperatorGenerator implements AdaptiveJoin {
+
+    private final int[] leftKeys;
+
+    private final int[] rightKeys;
+
+    private final FlinkJoinType joinType;
+
+    private final boolean[] filterNulls;
+
+    private final RowType leftType;
+
+    private final RowType rightType;
+
+    private final GeneratedJoinCondition condFunc;
+
+    private final int leftRowSize;
+
+    private final long leftRowCount;
+
+    private final int rightRowSize;
+
+    private final long rightRowCount;
+
+    private final boolean tryDistinctBuildRow;
+
+    private final long managedMemory;
+
+    private final OperatorType originalJoin;
+
+    private boolean leftIsBuild;
+
+    private boolean isBroadcastJoin;
+
+    public AdaptiveJoinOperatorGenerator(
+            int[] leftKeys,
+            int[] rightKeys,
+            FlinkJoinType joinType,
+            boolean[] filterNulls,
+            RowType leftType,
+            RowType rightType,
+            GeneratedJoinCondition condFunc,
+            int leftRowSize,
+            int rightRowSize,
+            long leftRowCount,
+            long rightRowCount,
+            boolean tryDistinctBuildRow,
+            long managedMemory,
+            OperatorType originalJoin) {
+        this.leftKeys = leftKeys;
+        this.rightKeys = rightKeys;
+        this.joinType = joinType;
+        this.filterNulls = filterNulls;
+        this.leftType = leftType;
+        this.rightType = rightType;
+        this.condFunc = condFunc;
+        this.leftRowSize = leftRowSize;
+        this.rightRowSize = rightRowSize;
+        this.leftRowCount = leftRowCount;
+        this.rightRowCount = rightRowCount;
+        this.tryDistinctBuildRow = tryDistinctBuildRow;
+        this.managedMemory = managedMemory;
+        this.originalJoin = originalJoin;
+    }
+
+    @Override
+    public StreamOperatorFactory<?> genOperatorFactory(
+            ClassLoader classLoader, ReadableConfig config) {
+        if (isBroadcastJoin || originalJoin == OperatorType.ShuffleHashJoin) {
+            return HashJoinOperatorUtil.generateOperatorFactory(
+                    leftKeys,
+                    rightKeys,
+                    joinType,
+                    filterNulls,
+                    leftType,
+                    rightType,
+                    condFunc,
+                    leftIsBuild,
+                    leftRowSize,
+                    rightRowSize,
+                    leftRowCount,
+                    rightRowCount,
+                    tryDistinctBuildRow,
+                    managedMemory,
+                    config,
+                    classLoader);
+        } else {
+            return SorMergeJoinOperatorUtil.generateOperatorFactory(
```

Comment:
Why do we convert it to a sort-merge-join if it was originally a shuffled-hash-join which cannot be converted into a broadcast-hash-join?

Reply:
> Why do we convert it to a sort-merge-join if it was originally a shuffled-hash-join which cannot be converted into a broadcast-hash-join?

No, perhaps it is necessary to add some annotations here for clarification. The overall logic can be divided into the following two situations:

1. When broadcast conditions are met, all join operators will be transformed into HashJoinOperator (broadcast).
2. If adaptive join is not satisfied, the shuffle hash join will still generate the corresponding HashJoinOperator, and Sort Merge Join will roll back to SortMergeJoinOperator.

Reply:
Yes this logic is straight forward. I guess I overlooked something, but I cannot remember now.